core/src/main/scala/org/apache/spark/sql/streaming/DirectKafkaStreamSource.scala

```diff
@@ -16,63 +16,135 @@
 
 class DirectKafkaStreamSource extends StreamPlanProvider with DataSourceRegister {
 
-  override def shortName(): String = SnappyContext.DIRECT_KAFKA_STREAM_SOURCE
+  override def shortName(): String = SnappyContext.KAFKA_STREAM_SOURCE
 
   override def createRelation(sqlContext: SQLContext,
-      options: Map[String, String],
-      schema: StructType): BaseRelation = {
+                              options: Map[String, String],
+                              schema: StructType): BaseRelation = {
     new DirectKafkaStreamRelation(sqlContext, options, schema)
   }
 }
 
 final class DirectKafkaStreamRelation(
-    @transient override val sqlContext: SQLContext,
-    opts: Map[String, String],
-    override val schema: StructType)
-    extends StreamBaseRelation(opts)
+                                       @transient override val sqlContext: SQLContext,
+                                       opts: Map[String, String],
+                                       override val schema: StructType)
+  extends StreamBaseRelation(opts)
     with Logging with Serializable {
 
-  val topicsSet = options("topics").split(",").toSet
+  val topics = options("subscribe").split(",").toList
   val kafkaParams: Map[String, String] = options.get("kafkaParams").map { t =>
     t.split(";").map { s =>
       val a = s.split("->")
       (a(0), a(1))
     }.toMap
   }.getOrElse(Map())
 
-  val K = options.getOrElse("K", "java.lang.String")
-  val V = options.getOrElse("V", "java.lang.String")
-  val KD = options.getOrElse("KD", "kafka.serializer.StringDecoder")
-  val VD = options.getOrElse("VD", "kafka.serializer.StringDecoder")
+  val preferredHosts = LocationStrategies.PreferConsistent
+  val startingOffsets = JsonUtils.partitionOffsets(options("startingOffsets"))
 
   override protected def createRowStream(): DStream[InternalRow] = {
-    val ck: ClassTag[Any] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(K))
-    val cv: ClassTag[Any] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(V))
-    val ckd: ClassTag[Decoder[Any]] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(KD))
-    val cvd: ClassTag[Decoder[Any]] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(VD))
-    KafkaUtils.createDirectStream[Any, Any, Decoder[Any], Decoder[Any]](context,
-      kafkaParams, topicsSet)(ck, cv, ckd, cvd).mapPartitions { iter =>
-      val encoder = RowEncoder(schema)
-      // need to call copy() below since there are builders at higher layers
-      // (e.g. normal Seq.map) that store the rows and encoder reuses buffer
-      iter.flatMap(p => rowConverter.toRows(p._2).iterator.map(
-        encoder.toRow(_).copy()))
-    }
+    val consumerStrategies = ConsumerStrategies
+      .Subscribe[String, String](topics, kafkaParams, startingOffsets)
+
+    val stream = KafkaUtils.createDirectStream[String, String](context,
+      preferredHosts, consumerStrategies)
```

Comment:
Always assuming string keys and values in the stream. Why? We should support any key/value types like before. Also the Kafka-0.8 version used key and value decoder classes too. Do you know what Kafka-0.10 version is using for decoding? We don't have a mechanism to register Decoder implementations from SQL so removing the decoders here would mean that custom classes requiring own decoding won't work anymore. How does Kafka 0.10 do key/value decoding in any case?